File: crypto/xeddsa/SignedPreKey.go
package xeddsa

import (
  "crypto/rand"
)

type SignedPreKey struct {
  PreKey KeyPair
  Signature [64]byte
}

func NewSignedPreKey(identityKey PrivateKey) (*SignedPreKey, error) {
  random := rand.Reader

  preKey, err := Generate(random)
  if err != nil {
    return nil, err
  }
  sig := identityKey.Sign(random, preKey.PublicKey.Encode())
  ret := &SignedPreKey{ *preKey, *sig}

  return ret, nil
}

File: android/core/src/main/java/id/ridon/ngobrel/core/SignedPreKey.java
package id.ridon.ngobrel.core;

import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SignatureException;

/**
 * This class represents signed pre-keys
 */
public class SignedPreKey {

  final SignedPreKeyPublic pub;
  final PrivateKey priv;

  public SignedPreKey(PrivateKey key) throws IllegalDataSizeException, NoSuchAlgorithmException, InvalidKeyException, SignatureException {
    KeyPair preKey = new KeyPair();

    Signature sig = key.sign(preKey.publicKey.encode());

    pub = new SignedPreKeyPublic(preKey.publicKey, sig);
    priv = preKey.privateKey;
  }

  public final SignedPreKeyPublic getPublic() {
    return pub;
  }

  public final PrivateKey getPrivateKey() {
    return priv;
  }
}
